# Problem 2

Need to find:

$ 
f[x_0] \\
f[x_1] \\
f[x_0,x_1]
$

So,

\begin{gather*}

f[x_0,x_1,x_2] = \frac{f[x_1,x_2] - f[x_0,x_1]}{x_2 - x_0} = \frac{50}{7} \\

70 - 7f[x_0,x_1] = 50(\frac{7}{10}) \\

-7f[x_0,x_1] = 35 \\

\implies f[x_0,x_1] = 5 \\\\

f[x_1,x_2] = \frac{f(x_2) - f(x_1)}{x_2 - x_1} = 10 \\

6 - f(x_1) = 10 (\frac{3}{10}) \\

\implies f(x_1) = 3 \\\\

f[x_0,x_1] = \frac{f[x_1]-f[x_0]}{x_1 - x_0} = 5 \\
f(x_1) - f(x_0) = 5(\frac{4}{10}) \\
3 - f(x_0) = 2 \\
\implies f(x_0) = 1

\end{gather*}

# Problem 3

Since we are given $\Delta P(0) = 24$, we know that for all $x$, $\Delta^4P(x) = 24$ because P(x) is a 4th degree polynomial, and all 4th degree differences are constant.

We also know that 
$
\\
\Delta^4 P(x) = \Delta^3 (\Delta P(x)) \\
\implies \Delta^3 (P(x+1) - P(x)) 
\\
\implies \Delta^3 P(x+1) - \Delta^3 P(x)
\\\\
\implies
\Delta^4 P(x) + \Delta^3 P(x) = \Delta^3 P(x+1)\\
$

Let us use induction and change 4 into $k$ and 3 into $k-1$:

$
\implies \Delta^k P(x) + \Delta^{k-1} P(x) = \Delta^{k-1} P(x+1)
$

Let's use the given info that $\Delta^4 P(0) = 24, \Delta^3 P(0) = 6, \Delta^2 P(0) = 0$ to calculate all x's on the 3rd order difference:

In [1]:
# import numpy
import numpy as np
from icecream import ic

# get rid of scientific notation (because scope is not big)
np.set_printoptions(suppress=True)

# make res of vals
# col 1 is x, col 2 is 3rd order forward difference
thirdDegrees = np.zeros([11,2])

# make vals on x be increasing
for i in range(0,11):
    thirdDegrees[i,0] = i

# 4th degree val
fourth = 24

# set first val
thirdDegrees[0,1] = 6

# fill third degree array
for i in range(1,11):
    thirdDegrees[i,1] = fourth + thirdDegrees[i-1,1]

ic(thirdDegrees)

# do the same for 2nd degree but use third degree vals now

secondDegrees = np.zeros([11,2])
for i in range(0,11):
    secondDegrees[i,0] = i

# fill 2nd degree 0 val = 0
# already filled as 0

# fill with formula
for i in range(1,11):
    secondDegrees[i,1] = int(thirdDegrees[i-1,1] + secondDegrees[i-1,1])

# print out result
ic(secondDegrees)


ic| thirdDegrees: array([[  0.,   6.],
                         [  1.,  30.],
                         [  2.,  54.],
                         [  3.,  78.],
                         [  4., 102.],
                         [  5., 126.],
                         [  6., 150.],
                         [  7., 174.],
                         [  8., 198.],
                         [  9., 222.],
                         [ 10., 246.]])
ic| secondDegrees: array([[   0.,    0.],
                          [   1.,    6.],
                          [   2.,   36.],
                          [   3.,   90.],
                          [   4.,  168.],
                          [   5.,  270.],
                          [   6.,  396.],
                          [   7.,  546.],
                          [   8.,  720.],
                          [   9.,  918.],
                          [  10., 1140.]])


array([[   0.,    0.],
       [   1.,    6.],
       [   2.,   36.],
       [   3.,   90.],
       [   4.,  168.],
       [   5.,  270.],
       [   6.,  396.],
       [   7.,  546.],
       [   8.,  720.],
       [   9.,  918.],
       [  10., 1140.]])

And so, based on the output, when $x = 10, \Delta^2 P(10) = 1140$.